src/qibo/backends/__init__.py

```diff
@@ -1,59 +1,71 @@
 import os
 from qibo import config
 from qibo.config import raise_error, log, warnings
-from qibo.backends.numpy import NumpyBackend
-from qibo.backends.tensorflow import TensorflowBackend
-
+from qibo.backends.numpy import NumpyDefaultEinsumBackend, NumpyMatmulEinsumBackend
+from qibo.backends.tensorflow import TensorflowCustomBackend, TensorflowDefaultEinsumBackend, TensorflowMatmulEinsumBackend
+
+
+AVAILABLE_BACKENDS = {
+    "custom": TensorflowCustomBackend,
+    "tensorflow": TensorflowCustomBackend,
+    "defaulteinsum": TensorflowDefaultEinsumBackend,
+    "matmuleinsum": TensorflowMatmulEinsumBackend,
+    "tensorflow_defaulteinsum": TensorflowDefaultEinsumBackend,
+    "tensorflow_matmuleinsum": TensorflowMatmulEinsumBackend,
+    "numpy": NumpyDefaultEinsumBackend,
+    "numpy_defaulteinsum": NumpyDefaultEinsumBackend,
+    "numpy_matmuleinsum": NumpyMatmulEinsumBackend
+}
```

Comment:
I am checking this PR, and I think we could be a little bit more specific for this dict.
What do you think if we include some docs so when you raise an error we can show some message like:
```
ValueError: Unknown backend *. Please select one of the available backends: 
- custom: uses precompiled primitives, fastest simulation engine...
- tensorflow: same as custom...
- defaulteinsum: uses tf primitives, and einsum, faster on GPU...
- matmuleinsum: ...
```

Reply:
Another motivation to do that is the follow: suppose I set `qibo.set_backend("numpy")`, then when calling `qibo.get_backend()` it prints `numpy_defaulteinsum`, so this may confuse the user, and in the worst case break some programmatic check.

Reply:
I updated the error when an invalid backend is given so that a list of the available backends is printed. You can check the error message by passing a wrong value on purpose (for example do `qibo.set_backend("test")`).

Regarding the second point, I agree that it may be a bit confusing that `numpy` and `numpy_defaulteinsum` are essentially the same backend (same for `custom` and `tensorflow`). The reason I did it this way is because I wanted to enable the user to do `qibo.set_backend("numpy")` as a shortcut for `qibo.set_backend("numpy_defaulteinsum")`. If you think it is confusing we can remove `numpy` and keep only the `numpy_defaulteinsum` option.